apps/vaporgui/VaporWidgets.cpp

```diff
@@ -145,6 +146,127 @@ double VDoubleSpinBox::GetValue() const {
     return _value;
 }
 
+//
+// ====================================
+//
+VSlider::VSlider( QWidget* parent, const std::string& label, float min, float max )
+       : VaporWidget( parent, label )
+{
+    _min = min;
+    _max = max;
+    VAssert( _max > _min );
+    _currentVal = (_min + _max) / 2.0f;
+    
+    _qslider = new QSlider( this );
+    _qslider->setOrientation( Qt::Horizontal );
+    /* QSlider will always have its range in integers from 0 to 100. */
+    _qslider->setMinimum( 0 );
+    _qslider->setMaximum( 100 );
+    connect( _qslider, SIGNAL( sliderReleased() ), this, SLOT( _respondQSliderReleased() ) );
+    connect( _qslider, SIGNAL( sliderMoved(int) ), this, SLOT( _respondQSliderMoved(int) ) );
+    _layout->addWidget( _qslider );
+        
+    _qedit = new QLineEdit( this );
+    connect( _qedit, SIGNAL( editingFinished() ), this, SLOT( _respondQLineEdit() ) );
+    _layout->addWidget( _qedit );
+
+    /* update widget display */
+    float perc = (_currentVal - _min) / (_max - _min) * 100.0f;
+    _qslider->setValue( std::lround( perc ) );
+    _qedit->setText( QString::number( _currentVal, 'f', 3 ) );
+}
+
+VSlider::~VSlider() {}
+
+void
+VSlider::SetRange( float min, float max )
+{
+    VAssert( min < max );
+    _min = min;
+    _max = max;
+
+    /* keep the old _currentVal if it's still within the range.
+       Otherwise, re-assign the middle point to _currentVal */
+    if( _currentVal < min ||  _currentVal > max )
+    {
+        _currentVal = (min + max) / 2.0f;
+        float perc = (_currentVal - _min) / (_max - _min) * 100.0f;
+        _qslider->setValue( std::lround( perc ) );
+        _qedit->setText( QString::number( _currentVal, 'f', 3 ) );
+    }
+}
+
+void
+VSlider::SetCurrentValue( float val )
+{
+    /* Only respond if val is within range */
+    if( val >= _min && val <= _max )
+    {
+        _currentVal = val;
+        float perc = (_currentVal - _min) / (_max - _min) * 100.0f;
+        _qslider->setValue( std::lround( perc ) );
+        _qedit->setText( QString::number( _currentVal, 'f', 3 ) );
+    }
+}
+
+float
+VSlider::GetCurrentValue() const
+{
+    return _currentVal;
+}
+
+void
+VSlider::_respondQSliderReleased()
+{
+    /* QSlider is always giving a valid value, so no need to validate range */
+    int newvalI = _qslider->value();
+    float perc  = (float)newvalI / 100.0f;
```

Comment:
Could we change this variable name to "precision"?

Reply:
I'll fix it in my next PR.